src/db/transformer.ts

```diff
@@ -0,0 +1,291 @@
+import {assert} from '../asserts';
+import {assertBTreeNode} from '../btree/node';
+import * as btree from '../btree/mod';
+import {
+  IndexChangeMeta,
+  Meta,
+  MetaTyped,
+  SnapshotMeta,
+  getRefs as getRefsFromCommitData,
+  assertCommitData,
+} from './commit';
+import type {Hash} from '../hash';
+import type * as db from './mod';
+import type * as dag from '../dag/mod';
+import type {ReadonlyJSONValue} from '../json';
+import {HashType} from './hash-type';
+import type {Value} from '../kv/store';
+
+export class Transformer {
+  readonly dagWrite: dag.Write;
+  private readonly _transforming: Map<Hash, Promise<Hash>> = new Map();
+
+  constructor(dagWrite: dag.Write) {
+    this.dagWrite = dagWrite;
+  }
+
+  private _withTransformingCache(
+    h: Hash,
+    f: () => Promise<Hash>,
+  ): Promise<Hash> {
+    const newHash = this._transforming.get(h);
+    if (newHash !== undefined) {
+      return newHash;
+    }
+
+    const p = f();
+    this._transforming.set(h, p);
+    return p;
+  }
+
+  private _transformCommitWithCache(
+    h: Hash,
+    hashType = HashType.RequireStrong,
+  ): Promise<Hash> {
+    return this._withTransformingCache(h, () =>
+      this.transformCommit(h, hashType),
+    );
+  }
+
+  async transformCommit(
+    h: Hash,
+    hashType = HashType.RequireStrong,
+  ): Promise<Hash> {
+    const chunk = await this.getChunk(h);
+    if (!chunk) {
+      if (hashType === HashType.AllowWeak) {
+        return h;
+      }
+      throw new Error(`Chunk ${h} not found`);
+    }
+    const {data} = chunk;
+    assertCommitData(data);
+
+    const newCommitData = await this._transformCommitData(data);
+    return this._maybWriteChunk(h, newCommitData, data, getRefsFromCommitData);
+  }
+
+  protected shouldForceWrite(_h: Hash): boolean {
+    return false;
+  }
+
+  protected getChunk(h: Hash): Promise<dag.Chunk | undefined> {
+    return this.dagWrite.getChunk(h);
+  }
+
+  private async _maybWriteChunk<D extends Value>(
+    h: Hash,
+    newData: D,
+    oldData: D,
+    getRefs: (data: D) => readonly Hash[],
+  ): Promise<Hash> {
+    if (newData !== oldData || this.shouldForceWrite(h)) {
+      const newChunk = this.dagWrite.createChunk(newData, getRefs(newData));
+      await this.dagWrite.putChunk(newChunk);
+      return newChunk.hash;
```

Comment:
To confirm my understanding, we expect the oldData (non-transformed chunk) to be collected via the DAG garbage collection mechanism, correct?

Reply:
yes